### PR TITLE
Pattern Matching and Using Variable Integration

### PR DESCRIPTION
**Pull request by fayrose**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -621,14 +624,27 @@ private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSy
             else if (node.UsingKeyword != default)
             {
                 kind = LocalDeclarationKind.UsingVariable;
-            }
+                BoundLocalDeclaration[] boundDeclarations = new BoundLocalDeclaration[variableCount];
+                int i = 0;
+                foreach (var variableDeclaratorSyntax in variableList)
+                {
+                    boundDeclarations[i++] = BindVariableDeclaration(kind, isVar, variableDeclaratorSyntax, typeSyntax, declType, alias, diagnostics);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >BindVariableDeclaration [](start = 45, length = 23)

We're calling `BindVariableDeclaration` twice on each `using` declaration - once here and again in `if` statement below. Is the second bind necessary?

Also, we're dropping `disposeMethod` and `iDisposableConversion` in the second bind when there are multiple variables. Is that intended?

Comment (fayrose) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: This PR is only supposed to take care of the single variable case, where you can see that it is passed on. Julian preferred I do the multiple variable in a different PR. Will fix double-call now.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -604,44 +604,124 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
 
         private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
         {
-            var typeSyntax = node.Declaration.Type.SkipRef(out _);
+            var variableList = node.Declaration.Variables;
+            int variableCount = variableList.Count;
+
             bool isConst = node.IsConst;
+            bool IsUsing = node.UsingKeyword != default;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >I [](start = 17, length = 1)

Local names should start with lowercase letter.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 3]
```diff
@@ -604,44 +604,124 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
+            Conversion iDisposableConversion = default;
+            MethodSymbol disposeMethod = default;
 
             LocalDeclarationKind kind = LocalDeclarationKind.RegularVariable;
             if (isConst)
             {
                 kind = LocalDeclarationKind.Constant;
             }
-            else if (node.UsingKeyword != default)
+            if (IsUsing)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >if [](start = 12, length = 2)

Please keep as `else if`.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 4]
```diff
@@ -604,44 +604,124 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
                 }
+            }
 
-                return new BoundMultipleLocalDeclarations(node, boundDeclarations.AsImmutableOrNull());
+            return new BoundUsingLocalDeclarations(node, disposeMethod, iDisposableConversion, declarations);
+        }
+
+        /// <summary>
+        /// Checks for a Dispose method on exprType in the case that there is no explicit
+        /// IDisposable conversion.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Consider removing this part of the comment since this method does not check any conversion.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 5]
```diff
@@ -604,44 +604,124 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
-            int variableCount = variableList.Count;
+        internal BoundMultipleLocalDeclarations BindUsingVariableDeclaration(Binder originalBinder,
+                                                                             DiagnosticBag diagnostics,
+                                                                             bool hasErrors,
+                                                                             SyntaxNode node,
+                                                                             VariableDeclarationSyntax declarationSyntax,
+                                                                             out Conversion iDisposableConversion,
+                                                                             out MethodSymbol disposeMethod)
+        {
+            TypeSymbol iDisposable = this.Compilation.GetSpecialType(SpecialType.System_IDisposable);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >TypeSymbol iDisposable [](start = 12, length = 22)

Declaration can be moved inside `else`.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 6]
```diff
@@ -604,44 +604,124 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
+                                                    diagnostics.HasAnyErrors(),
+                                                    node,
+                                                    node.Declaration,
+                                                    out iDisposableConversion,
+                                                    out disposeMethod);
+                
+            }
+            else
+            {
+                if (IsMultiple)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Is there a reason to change this from the original?
```
if (variableCount == 1)
{
    return BindVariableDeclaration(...);
}
else
{
    ...
    return new BoundMultipleLocalDeclarations(...);
}
```
If not, consider reverting to avoid unnecessary changes.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 3 above)
Comment (fayrose) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I can't, the else on line 638 needs to catch all cases, including ones within the isConst on line 620.

Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: If the `else` at line 637 is removed, and the `else if` at line 623 is restored, the original structure is preserved.

---
In reply to: [206625160](https://github.com/dotnet/roslyn/pull/28920#discussion_r206625160) [](ancestors = 206625160)

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 7]
```diff
@@ -604,44 +604,120 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
 
             LocalDeclarationKind kind = LocalDeclarationKind.RegularVariable;
             if (isConst)
             {
                 kind = LocalDeclarationKind.Constant;
             }
-            else if (node.UsingKeyword != default)
+            if (node.UsingKeyword != default)
             {
                 kind = LocalDeclarationKind.UsingVariable;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >kind = LocalDeclarationKind.UsingVariable [](start = 16, length = 41)

Not used.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 8]
```diff
@@ -604,44 +604,120 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
 
+            bool isConst = node.IsConst;
+            var typeSyntax = node.Declaration.Type.SkipRef(out _);
             bool isVar;
             AliasSymbol alias;
             TypeSymbol declType = BindVariableType(node.Declaration, diagnostics, typeSyntax, ref isConst, isVar: out isVar, alias: out alias);
 
-            // UNDONE: "possible expression" feature for IDE
+            Conversion iDisposableConversion = default;
+            MethodSymbol disposeMethod = default;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: These declarations can be moved inside the `if (node.UsingKeyword != default)` block, and the initializers are not needed.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenUsingVariableTests.cs [hunk 9]
```diff
@@ -345,6 +345,58 @@ .maxstack  1
...
+  {
+    IL_000e:  ldloc.0
+    IL_000f:  brfalse.s  IL_0017
+    IL_0011:  ldloc.0
+    IL_0012:  callvirt   ""void C1.Dispose()""
+    IL_0017:  endfinally
+  }
+  IL_0018:  ret
 }");
         }
```
[the remark this answers is not in the document]
Comment (fayrose) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenUsingVariableTests.cs]: Added

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 10]
```diff
@@ -604,44 +604,118 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
+                {
+                    return BindVariableDeclaration(kind, isVar, variableList[0], typeSyntax, declType, alias, diagnostics, node);
+                }
             }
+        }
 
-            var variableList = node.Declaration.Variables;
-            int variableCount = variableList.Count;
+        internal ImmutableArray<BoundLocalDeclaration> BindUsingVariableDeclaration(Binder originalBinder,
+                                                                             DiagnosticBag diagnostics,
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Indentation is off.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 11]
```diff
@@ -604,44 +604,118 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
 
         private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
         {
-            var typeSyntax = node.Declaration.Type.SkipRef(out _);
-            bool isConst = node.IsConst;
+            var variableList = node.Declaration.Variables;
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: These look like they can stay in the `else` below, since they're not used in the `using var` form.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 12]
```diff
@@ -841,11 +919,14 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
             TypeSymbol declTypeOpt,
             AliasSymbol aliasOpt,
             DiagnosticBag diagnostics,
-            CSharpSyntaxNode associatedSyntaxNode = null)
+            CSharpSyntaxNode associatedSyntaxNode = null,
+            MethodSymbol disposeMethod = default,
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: The new parameters are unused now.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 13]
```diff
@@ -817,7 +891,9 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
...
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: You shouldn't need the dispose or conversion after you remove them from `BindVariableDeclaration`

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 14]
```diff
@@ -2304,7 +2385,8 @@ internal virtual BoundForStatement BindForParts(DiagnosticBag diagnostics, Binde
             return this.Next.BindForParts(diagnostics, originalBinder);
         }
 
-        internal BoundStatement BindForOrUsingOrFixedDeclarations(VariableDeclarationSyntax nodeOpt, LocalDeclarationKind localKind, DiagnosticBag diagnostics, out ImmutableArray<BoundLocalDeclaration> declarations)
+        internal BoundStatement BindForOrUsingOrFixedDeclarations(VariableDeclarationSyntax nodeOpt, LocalDeclarationKind localKind, DiagnosticBag diagnostics, out ImmutableArray<BoundLocalDeclaration> declarations,
+                                                                  Conversion iDisposableConversion = default, MethodSymbol disposeMethod = default)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: The conversion and dispose method parameters aren't needed now.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 15]
```diff
@@ -604,44 +604,119 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
 
         private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
         {
-            var typeSyntax = node.Declaration.Type.SkipRef(out _);
+            
             bool isConst = node.IsConst;
-
+            var typeSyntax = node.Declaration.Type.SkipRef(out _);
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: This change doesn't look necessary.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 16]
```diff
@@ -604,44 +604,119 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
+                                                    node.Declaration,
+                                                    out iDisposableConversion,
+                                                    out disposeMethod);
+                return new BoundUsingLocalDeclarations(node, disposeMethod, iDisposableConversion, declarations);
             }
-            else if (node.UsingKeyword != default)
+            else
             {
-                kind = LocalDeclarationKind.UsingVariable;
+                LocalDeclarationKind kind = LocalDeclarationKind.RegularVariable;
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: You could probably write this

`var kind = isConst ? LocalDeclarationKind.Constant : LocalDeclarationKind.RegularVariable;`

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 17]
```diff
@@ -604,44 +604,119 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
+                var variableList = node.Declaration.Variables;
+                int variableCount = variableList.Count;
+
+                if (isConst)
+                {
+                    kind = LocalDeclarationKind.Constant;
+                }
+                if (variableCount > 1)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: It looks like this code reverses the logic from the previous version, but I don't see why.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 18]
```diff
@@ -604,44 +604,119 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
...
+        }
 
-                return new BoundMultipleLocalDeclarations(node, boundDeclarations.AsImmutableOrNull());
+        /// <summary>
+        /// Checks for a Dispose method on exprType in the case that there is no explicit
+        /// IDisposable conversion.
+        /// </summary>
+        /// <param name="exprType">Type of the expression over which to iterate</param>
+        /// <param name="diagnostics">Populated with warnings if there are near misses</param>
+        /// <returns>True if a matching method is found with correct return type.</returns>
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: This comment looks wrong.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 19]
```diff
@@ -829,7 +904,7 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
                                            declTypeOpt,
                                            aliasOpt,
                                            diagnostics,
-                                           associatedSyntaxNode);
+                                           associatedSyntaxNode,);
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Syntax error

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 20]
```diff
@@ -2342,7 +2417,7 @@ internal BoundStatement BindForOrUsingOrFixedDeclarations(VariableDeclarationSyn
             for (int i = 0; i < count; i++)
             {
                 var variableDeclarator = variables[i];
-                var declaration = BindVariableDeclaration(localKind, isVar, variableDeclarator, typeSyntax, declType, alias, diagnostics);
+                var declaration = BindVariableDeclaration(localKind, isVar, variableDeclarator, typeSyntax, declType, alias, diagnostics, disposeMethod: disposeMethod, iDisposableConversion: iDisposableConversion);
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Syntax error.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 21]
```diff
@@ -735,7 +735,12 @@
   <Node Name="BoundMultipleLocalDeclarations" Base="BoundStatement">
     <Field Name="LocalDeclarations" Type="ImmutableArray&lt;BoundLocalDeclaration&gt;"/>
   </Node>
- 
+
+  <Node Name="BoundUsingLocalDeclarations" Base="BoundMultipleLocalDeclarations">
+    <Field Name="DisposeMethodOpt" Type="MethodSymbol" Null="Allow"/>
+    <Field Name="IDisposableConversion" Type="Conversion?" Null="Allow"/>
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: I originally suggested using a nullable here, but it looks like most other places just use Conversion.NoConversion, so we should probably not make this nullable.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 22]
```diff
@@ -604,44 +604,114 @@ internal BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementS
 
         private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
         {
-            var typeSyntax = node.Declaration.Type.SkipRef(out _);
             bool isConst = node.IsConst;
-
+            var typeSyntax = node.Declaration.Type.SkipRef(out _);
```
[the remark this answers is not in the document]
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Agree this should be removed. 

---
In reply to: [207031922](https://github.com/dotnet/roslyn/pull/28920#discussion_r207031922) [](ancestors = 207031922)
